Add new Slack handlers using Slackbot and webhooks

hkdobrev opened the pull request:
See https://github.com/Seldaek/monolog/issues/743#issuecomment-242994019

This is keeping the previous `SlackHandler`, but extracts the message formatting logic and introduces two new Slack handlers:
- `SlackbotHandler` - this is the simplest way to log to Slack and most people could use that. It doesn't allow fancy formatting, but has autolinking and the simplest configuration with two clicks in the Slack settings and getting a token.
- `SlackWebhookHandler` - This is a bit more advanced as it allows for the same formatting and adding attachements as the `SlackHandler`, but it uses the Slack Webhooks which are quite easier to set up than the OAuth API. The minimum it requires is a webhook URL which is a retrievable with about two clicks in the Slack settings. Even the channel is optional as it could be configured in the integration.

This is probably not the leanest way to implement what we've discussed in https://github.com/Seldaek/monolog/issues/743, but after working at this I believe this is the simplest and the right approach.

While we could implement everything in one handler the API would become horrible and it would be harder to understand from users. Providing choice with simple to use separate classes is better IMHO. Apart from this the `SlackHandler` extends the `SocketHandler` and the webhooks and the Slackbot wouldn't work nicely with that.

What do you think?

@Seldaek @gkedzierski @Gummibeer @nisbeti

---
- [x] Extracting Slack formatting logic in `SlackRecord`
- [x] `SlackbotHandler`
- [x] `SlackWebhookHandler`
- [x] Refactor `SlackHandler` using the new `SlackRecord`
- [x] Add new handlers to the docs
- [x] Unit tests of `SlackRecord`
- [x] Unit tests for `SlackbotHandler`
- [x] Unit tests for `SlackWebhookHandler`
- [x] Update unit tests for `SlackHandler`

Gummibeer commented:
I will Test it tomorrow and give you Feedback.

Seldaek commented:
This seems fine to me implementation wise, if you can write tests for SlackRecord at least that'd be nice, testing the curl calls isn't really easy nor worth it IMO.

I'd also be happy to get feedback from people actually using this.. :)

Seldaek commented:
BTW it might be worth looking at this at the same time if the Slack stuff gets a rewrite.. https://github.com/Seldaek/monolog/issues/745

And also marking the original SlackHandler as deprecated would be good I think (I will remove it from master/2.0 then), or do you think there is really a point in having 3?

greeny commented:
I know that this might be a case for all handlers, not just this one, but what if I have disabled curl extension? There is also way to send requests through `file_get_contents` (you can check my implementation in https://github.com/greeny/nette-slack-logger/blob/master/src/SlackLogger.php#L77). Are we going to support this?

hkdobrev commented:
@Seldaek I've started working on the tests and while doing so I've deprecated a couple of protected methods in `SlackHandler` which are better accessed through `SlackRecord`. However I don't see why we'd deprecate the `SlackHandler`. It's the only handler using the Slack API directly.

@greeny Could you point me to Monolog handlers which employ either `cURL` and if not available fall back to `file_get_contents()`? I don't think using only `file_put_contents()` is nice. I guess we could use Guzzle instead which would do the check and the fallback for us.

greeny commented:
@hkdobrev I did not check monolog for handlers, but I think many of them are firing requests through curl. Guzzle is fine solution, I was just asking, if servers without curl can use this handler too.

<offtopic>Regarding my implementation, that was first only internal service which i published later.</offtopic>

Seldaek commented:
Most handlers use curl or raw sockets yes. I think that's fine to be honest, it's the first time anyone even asks about it in all these years. I wouldn't add a dependency on guzzle just for that.

greeny commented:
@Seldaek it is not mine case (i have curl everywhere), but I know some hostings, that have curl disabled. Thats why I have `file_get_contents` in my implementation.

But I agree curl is very common nowdays.

@hkdobrev what about support for non-emoji icons? (`"icon_url": "https://slack.com/img/icons/app-57.png"`)

stof commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -0,0 +1,224 @@
+                        $attachment['fields'][] = array(
+                            'title' => "Extra",
+                            'value' => $this->stringify($record['extra']),
+                            'short' => $this->useShortAttachment,
```
could be ``true`` here, as we know the value. It may be more readable

stof commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -0,0 +1,224 @@
+                            $attachment['fields'][] = array(
+                                'title' => $var,
+                                'value' => $val,
+                                'short' => $this->useShortAttachment,
```
could be ``false`` directly

stof commented on src/Monolog/Handler/SlackHandler.php:
```diff
@@ -81,25 +46,36 @@ class SlackHandler extends SocketHandler
         if (!extension_loaded('openssl')) {
-            throw new MissingExtensionException('The OpenSSL PHP extension is required to use the SlackHandler');
+            throw new MissingExtensionException(
+                'The OpenSSL PHP extension is required to use the SlackHandler'
```
this change should be reverted

stof commented on src/Monolog/Handler/SlackHandler.php:
```diff
@@ -81,25 +46,36 @@ class SlackHandler extends SocketHandler
+        $level = Logger::CRITICAL,
+        $bubble = true,
+        $useShortAttachment = false,
+        $includeContextAndExtra = false
```
Monolog always has the signature on one line

stof commented on src/Monolog/Handler/SlackbotHandler.php:
```diff
@@ -0,0 +1,84 @@
+    private $channel;
+
+    /**
+     * @param  string $token   Slackbot token
```
missing phpdoc for ``$slackTeam``

stof commented on src/Monolog/Handler/SlackWebhookHandler.php:
```diff
@@ -0,0 +1,96 @@
+
+    /**
+     * @param  string      $webhookUrl             Slack Webhook URL
+     * @param  string      $channel                Slack channel (encoded ID or name)
```
``string|null``

martijncalker commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -0,0 +1,224 @@
+                }
+            }
+
+            $dataArray['attachments'] = json_encode(array($attachment));
```
Not sure if I use it the right way, but now I've got:
```json
{"username":"TestBot","text":"","attachments":"[{\\"fallback\\":\\"Testing Slackhandler!\\",\\"color\\":\\"good\\",\\"fields\\":[{\\"title\\":\\"Level\\",\\"value\\":\\"INFO\\",\\"short\\":true}],\\"title\\":\\"Message\\",\\"text\\":\\"Testing Slackhandler!\\"}]","channel":"development-logging"}
```

Changing this line to:
```php
$dataArray['attachments'] = [$attachment];
```
Output:
```json
{"username":"TestBot","text":"","attachments":[{"fallback":"Testing Slackhandler!","color":"good","fields":[{"title":"Level","value":"INFO","short":true}],"title":"Message","text":"Testing Slackhandler!"}],"channel":"development-logging"}
```

hkdobrev replied:
@martijncalker Thanks for your feedback! This is now fixed thanks to the work of @an1zhegorodov

an1zhegorodov commented:
@Seldaek 
> I'd also be happy to get feedback from people actually using this.. :)

Happy to see webhooks implementation, because it's very simple to configure.

Using webhooks and curl as a temporary solution. Tried using current SlackHandler, but it silently failed to send messages. The weird thing though, it worked like a charm when I was xdebugging. Could be env issue on my side, anyone faced this behavior?

Seldaek commented:
@hkdobrev any update here? Should I wait a couple days to get a chance to include this in the next 1.x release or will you need more time to finalize it? No stress I just need to know :)

an1zhegorodov commented:
@Seldaek Wait couple of days pls. I'm using custom handler on my project, but would love this implementation. I've just finished tests for SlackRecord, planning to fix tests for SlackHandler and fix thing related to @stof comments.

an1zhegorodov commented:
@Seldaek [done](https://github.com/hkdobrev/monolog/pull/1). Now waiting for feedback from @hkdobrev

hkdobrev commented:
@an1zhegorodov Thank you for your work! `SlackRecord` wasn't really my code as I just extracted it from the previous `SlackHandler` and I wasn't really keen on refactoring it. I have now cherry-picked your commits and fixing conflicts in the meantime as I had some things locally which I haven't pushed. For example I've moved colours to constants, deprecated old methods and exposed the `SlackRecord` instance instead. I've started with a couple of tests and I've merged our `SlackRecordTest`.

Could you please confirm that your changes didn't incur any backwards incompatible changes? Not just to the PHP API, but also to the way the pre-existing `SlackHandler` is configured?

hkdobrev commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -115,59 +115,36 @@ public function getSlackData(array $record)
-                    'value' => $record['level_name'],
-                    'short' => true,
-                );
+                $attachment['fields'][] = $this->generateAttachmentField('Level', $record['level_name'], true);
```
I'm sorry for commenting on an outdated diff, will copy my comment on the updated line as it is still valid and so the comment could be visible.

hkdobrev commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -0,0 +1,249 @@
+                $attachment['title'] = $record['level_name'];
+            } else {
+                $attachment['title'] = 'Message';
+                $attachment['fields'][] = $this->generateAttachmentField('Level', $record['level_name'], true);
```
@an1zhegorodov I think here in the `else` of `if ($this->useShortAttachment)` we should really pass `false` for the short attachment option instead of `true`, don't you think? Just like @stof's comment: https://github.com/Seldaek/monolog/pull/846#discussion_r83167917

@stof While I agree with your comment and would have changed it without thinking twice, this is an example where not using the property introduces a bug more easily.

an1zhegorodov replied:
@hkdobrev It's not about `useShortAttachment` in this case. Check out [current](https://github.com/Seldaek/monolog/blob/0fb49dcfdf7318c0bb01bf6876a44bac0a4a99d1/src/Monolog/Handler/SlackHandler.php#L170) implementation. It also has `'short' => true`. That's because `level_name` is really short: `INFO`, `DEBUG`, `ERROR`.

hkdobrev replied:
@an1zhegorodov Thanks for pointing this out! I got confused from the if-else and for being away from this for a while.

hkdobrev commented:
@greeny 

> @hkdobrev what about support for non-emoji icons? (`"icon_url": "https://slack.com/img/icons/app-57.png"`)

This is a good suggestion. However, I wanted to limit the scope of this PR to provide the current functionality, but through the Slackbot or Webhooks without going through the Slack API. I think it should be extended to support everything Slack supports. But this could be done in a later Monolog release, ideally in 2.x when it could be refactored with breaking the public interface of the handler and its configuration so it could have a more flexible way to support everything Slack offers and even future features of Slack webhooks without changing the API and addition additional arguments to methods.

What do you think? Do you think this should be done as part of this PR? Cheers!

hkdobrev commented:
@Seldaek 

> And also marking the original `SlackHandler` as deprecated would be good I think (I will remove it from master/2.0 then), or do you think there is really a point in having 3?

I've marked a couple of methods as deprecated now. However, I disagree with marking the whole handler as deprecated as people who already use the Slack API could have a valid case and could continue to use it. Using the Slack API is not deprecated, so I don't see why we should deprecated the handler. Although, 2.x would be a good place to refactor the public interface and allow for more flexibility so the interface doesn't depend on every property Slack supports, I don't see why to deprecate the whole method of logging to the Slack API.

What do you think? Are you still considering removing the whole handler in 2.x?

hkdobrev commented:
Please let me know if you'd like commits squashed or partially squashed.

Thanks to all for all the feedback and help! I'm sorry for delaying it!

an1zhegorodov commented:
> Could you please confirm that your changes didn't incur any backwards incompatible changes? Not just to the PHP API, but also to the way the pre-existing SlackHandler is configured?

@hkdobrev I didn't change public API, only made `SlackHandler` constructor one line. Any hints what I should look for? I think it's worth to try it out locally on `1.x`, what do you think?

an1zhegorodov commented on src/Monolog/Handler/Slack/SlackRecord.php:
```diff
@@ -0,0 +1,249 @@
+                }
+            }
+
+            $dataArray['attachments'] = array($attachment);
```
Removed `json_encode` here. I think this will break `SlackHandler`, despite working fine with `SlackWebhookHandler`. `SlackHandler` sends data with `application/x-www-form-urlencoded` headers, and it looks like slack expects `attachments` field to be json_encoded in that case.

A few optoins I figured:
- json_encode `attachments` in `SlackHandler`
- change content-type to `application/json` and json_encode the whole record in `SlackHandler`

I like second more. What do you think @Seldaek @hkdobrev ?

hkdobrev replied:
@an1zhegorodov Could you elaborate more on the second option? Would that work with the Slack API?

an1zhegorodov replied:
@hkdobrev That should work according to [Slack API documentation](https://api.slack.com/methods/chat.postMessage)

> invalid_post_type	
> The method was called via a POST request, but the specified Content-Type was invalid. Valid types are: application/json application/x-www-form-urlencoded multipart/form-data text/plain.

I'll work on this tonight and test it out. Sorry for missing that in the first place.

hkdobrev replied:
@an1zhegorodov That may be an outdated answer, though.

1. It's from 2015.
2. The Slack API documentation you have linked earlier says JSON is supported.
3. Slack API website now has a message builder tool and it uses JSON exclusively. See [an example](https://api.slack.com/docs/messages/builder?msg=%7B%22attachments%22%3A%5B%7B%22fallback%22%3A%22Required%20plain-text%20summary%20of%20the%20attachment.%22%2C%22color%22%3A%22%2336a64f%22%2C%22pretext%22%3A%22Optional%20text%20that%20appears%20above%20the%20attachment%20block%22%2C%22author_name%22%3A%22Bobby%20Tables%22%2C%22author_link%22%3A%22http%3A%2F%2Fflickr.com%2Fbobby%2F%22%2C%22author_icon%22%3A%22http%3A%2F%2Fflickr.com%2Ficons%2Fbobby.jpg%22%2C%22title%22%3A%22Slack%20API%20Documentation%22%2C%22title_link%22%3A%22https%3A%2F%2Fapi.slack.com%2F%22%2C%22text%22%3A%22Optional%20text%20that%20appears%20within%20the%20attachment%22%2C%22fields%22%3A%5B%7B%22title%22%3A%22Priority%22%2C%22value%22%3A%22High%22%2C%22short%22%3Afalse%7D%5D%2C%22image_url%22%3A%22http%3A%2F%2Fmy-website.com%2Fpath%2Fto%2Fimage.jpg%22%2C%22thumb_url%22%3A%22http%3A%2F%2Fexample.com%2Fpath%2Fto%2Fthumb.png%22%2C%22footer%22%3A%22Slack%20API%22%2C%22footer_icon%22%3A%22https%3A%2F%2Fplatform.slack-edge.com%2Fimg%2Fdefault_application_icon.png%22%2C%22ts%22%3A123456789%7D%5D%7D).

Have you tried it out on the API? I might be able to try it out tomorrow if you won't be able to.

an1zhegorodov commented on src/Monolog/Handler/SlackWebhookHandler.php:
```diff
@@ -0,0 +1,101 @@
+     * @param  int         $level                  The minimum logging level at which this handler will be triggered
+     * @param  bool        $bubble                 Whether the messages that are handled can bubble up the stack or not
+     */
+    public function __construct(
```
@hkdobrev Are we going to one-line arguments for consistency?
Also, we can make the order of arguments the same as in `SlackHandler` (except here we'll have `webhookUrl` instead of `token`). This makes switching between `SlackHandler` and `SlackWebhookHandler` much easier.

an1zhegorodov commented on src/Monolog/Handler/SlackHandler.php:
```diff
@@ -260,36 +151,40 @@ protected function write(array $record)
+            E_USER_DEPRECATED
+        );
+
+        return $this->slackRecord->getAttachmentColor($level);
```
Just a thought. What if someone extended `SlackHanlder` just to make colors mapping different? He'll get default colors mapping after update, right? Looks like we're breaking backwards compatibility here. Same happens with `stringify`.

hkdobrev replied:
Yes. We would be breaking extensions which adhere to the Liskov principle, because we are not calling this method anymore. The deprecation notice would help, but not to people not familiar with Monolog. However, this is only about people extending `SlackHandler` as we are not changing anything for people using the public API. This is only the `protected` API. So I think we could close our eyes for that BC change here, because we have a deprecation notice for people who extended it.

@Seldaek Do you think it's OK breaking the use of classes extended `SlackHandler`?

Seldaek replied:
Yeah I think I can live with a minor break there.

Seldaek commented:
Note that the 2.0 constructor arguments are still very likely to change 
so don't attach too much importance to that.. Make it work for 1.x and 
we'll see.

an1zhegorodov commented:
@Seldaek Done that yesterday. @hkdobrev needs some time to try it out.

I'm using `SlackHandler` at work and have some improvements ideas, but I guess that'll be a separate PR.

mirfilip commented:
I haven't looked into the code yet, but I have a question. I'm currently using [this SlackWebhookHandler](https://github.com/Pageon/SlackWebhookMonolog). The one thing which is missing is the good formatter for `$context` and `$extra`. Yes, it's readable for the message, but when you pass context/extra, there is no good formatter available.

The formatter that would pretty print the array would be preferred. Back when my colleagues and I used our own logger, we used to turn on the markdown option (`mrkdwn`) and add an attachment with json serialized text. That's the prettiest way we come up with.

Any solutions to this problem here?

an1zhegorodov commented:
@mirfilip I guess the first thing is that `extra` and `context` should not really be included to the message, since we're adding them as attachment fields.

As for me, it requires much more work in terms of formatting, e.g. the user should be able to configure it to hide/show fields in context/extra.

Can you send us some screenshots of what you have in Slack using `SlackWebhookHandler`?

hkdobrev commented:
@Seldaek I've added some very basic unit tests for `SlackbotHandler` and `SlackWebhookHandler` as well. I know they look a bit like dummy tests, but I believe there should be at least one unit test in the test suite which would actually instantiate a handler. This would check for things like syntax errors and basic omissions.

With that I think this PR is ready. Thanks to @an1zhegorodov for picking up the slack (pun intended)! 🎉  💪

mirfilip commented:
@an1zhegorodov Stringify is indeed _some_ improvement, but with multidimensional arrays it's still pretty much useless.

One way I've seen this solved is either:
- Normalize (`NormalizerFormatter`) first, then json encode with pretty print
or 
-  `print_r` or similar

an1zhegorodov commented:
@mirfilip Once this is merged, I'll come up with a separate PR addressing this and a few other issues. Would appreciate your feedback.

hkdobrev commented:
@mirfilip @an1zhegorodov I also think it's better a new issue to be opened for this and tackled separately from the majority of the Slack handlers work.

wysow commented:
Hey guys, Thanks for this amazing work, can we see it merged soon @Seldaek ?

Seldaek commented:
Thanks everyone :)

Seldaek commented:
1.22 is out..

gmponos commented:
May I ask why cURL was preferred over Socket? Is it possible you could to move to socket instead of cURL?

hkdobrev commented:
The `SlackHandler` uses a socket with the streaming API. The simpler `SlackbotHandler` and the `SlackWebhookHandler` use cURL.

I'm not sure if these handlers could use a socket. cURL is simpler and based on their examples in the docs. I think only the streaming API would require a socket.

Why do you need to use a socket? If you use webhooks or the Slackbot, their examples give you just a webhook URL.

gmponos commented:
I hope I am not saying something incredibly stupid but I think using Sockets you can fire and forget am HTTP request from PHP.

https://stackoverflow.com/a/14668762/4158811
https://stackoverflow.com/a/14982587/4158811

The reason is that I don't want my application to break/stay halt if my Slack service is down.

I can see also that with the current implementation of SlackWebhook it retries the cURL request five times and I don't see a timeout setting so most probably my app will stay frozen while trying to call the webhook.

Not sure though if SocketHandler of monolog is also waiting for the response or it has a "fire and forget" logic.

mintobit commented:
"Fire and forget" would only work if there's no need to write data to the socket. We need to do that, so it would be locking operation with either Sockets or cURL.
To avoid application crash if Slack is down, you can use [WhatFailureGroupHandler](https://github.com/Seldaek/monolog/blob/master/doc/02-handlers-formatters-processors.md#wrappers--special-handlers).
Timeout setting is indeed missing, can be easily implemented the same way as in `SocketHandler` with fallback to ini settings.

If performance is critical, I'd also consider using `AmqpHandler` or `RedisHandler` and then sending messages to Slack using separate process.
